### ISSUE TITLE
Should we allow breaking changes to crate interfaces outside of major version bumps

### ISSUE THREAD

**Issue by willhickey** [anza-xyz/agave#3724]
[PR 3377](https://github.com/anza-xyz/agave/pull/3377) has ignited a philosophical debate which warrants public input.

## Context
The PR makes small breaking changes to the public interface on crates. The crates in question were not really intended to be a public interface, but we know they are being used directly outside of the Agave codebase.

## The argument against PR 3377:
The interfaces are public, regardless of our preconceptions about which ones should be used. Since we're no longer in beta (v0.x) we should only break public interfaces with major version bumps. Changes like this should be rolled out with new functions that wrap the old ones and deprecations on the old functions.

## The argument in favor of PR 3377:
We should only guarantee the interfaces on crates that we intended for public use. Developers interfacing directly with other crates can do so at their own risk. Introducing friction for changing interfaces reduces the likelihood that Agave developers will take the time to clean up existing code (eg functions with `#[allow(clippy::too_many_arguments)]`)

**Comment by t-nelson** [comment]
how this should _probably_ work is that we featurize everything non-public under `agave-internal`, make that feature non-default, then anything we're comfortable adhering to semver with moves out from under the feature

**Comment by steviez** [comment]
> how this should probably work is that we featurize everything non-public under agave-internal, make that feature non-default, 

I like this idea and am in favor of it. We want separate crates within our repo for better organization and modularization, but this stuff really shouldn't be public. So, `pub(crate)` is not an option for us for cross-crate deps (ie `core` and `runtime`)

> then anything we're comfortable adhering to semver with moves out from under the feature

Along this line, I think it would be interesting to hear what teams are using. We can certainly hear them out if they want us to move things out of internal feature (or really just have stable interface), but I think we need to protect our ability to clean things up and iterate quickly too

**Comment by steviez** [comment]
On a similar line of thought, once `solana-sdk` is evicted from the monorepo, I think it makes sense to still have an `agave-sdk` or `agave-internal-sdk` crate. A place for stuff like `Packet`, things common between `ledger-tool` and `validator`, etc to live

**Comment by t-nelson** [comment]
> On a similar line of thought, once `solana-sdk` is evicted from the monorepo, I think it makes sense to still have an `agave-sdk` or `agave-internal-sdk` crate. A place for stuff like `Packet`, things common between `ledger-tool` and `validator`, etc to live

junk drawer libraries are an anti-pattern

**Comment by steviez** [comment]
> junk drawer libraries are an anti-pattern

The status quo is this random stuff scattered across `sdk`, `ledger`, `core`, `runtime` and who knows where else. Breaking each individual item into its' own crate is an option, but I'd be curious if you or anyone else reading this have any alternatives in mind

**Comment by t-nelson** [comment]
it's simple. each crate that has public api has it's own sdk module where the public interface lives. everything else is private

`solana-core` is also a junk drawer library <norman-rockwell.jpg>

**Comment by nickfrosty** [comment]
anything that is a breaking change should follow semver, that way ecosystem developers can get a consistent experience across the solana stack (like any dev would expect with any other tool). the solana ecosystem and the agave client is mature enough for this. lets stop shooting application layer devs in the foot

intended or not, some things are public and being used by the ecosystem. even if something was not _meant_ to be public and used, it should **only** be removed (or made private) in a major version bump. if its public, some dev is bound to use it (like it or not). the only clean way to remove it (or make not public) is in a major version bump

> it's simple. each crate that has public api has it's own sdk module where the public interface lives. everything else is private

wholeheartedly agree with @t-nelson here! it also helps keep iteration nimble like @steviez (and im sure the rest of the engineers) want. reduce the public surface area for any random person to use (because if its available, someone will inevitably use it for some reason). cut a new major version of each crate and make only the minimally desired things public. then continuing to follow semver becomes really easy

> @steviez - we need to protect our ability to clean things up and iterate quickly too

if the public surface area is reduced, then it should help keep iteration nimble. if its public, expect at least someone to use it for some random reason. therefore must follow semver

**Comment by steviez** [comment]
>> The status quo is this random stuff scattered across sdk, ledger, core, runtime and who knows where else. Breaking each individual item into its' own crate is an option, but I'd be curious if you or anyone else reading this have any alternatives in mind
> 
> it's simple. each crate that has public api has it's own sdk module where the public interface lives. everything else is private

Either I'm missing something in your response or I don't think you answered my question. Let's consider the type `Packet` as a concrete example:
- This type is currently in `solana-sdk` (`sdk/src/packet.rs`) but probably shouldn't be
- This type is used by `solana-gossip`, `solana-turbine`, `solana-core`, `solana-perf` and probably some others. It doesn't really matter exactly how many, just that it is more than 1 crate

Where does the type `Packet` live; in a solo crate ?

**Comment by steveluscher** [comment]
> Where does the type Packet live; in a solo crate ?

This is spiritually similar to what we do on the JavaScript side; types like `Lamports` and `Commitment` live in their very own [npm package](https://github.com/anza-xyz/solana-web3.js/tree/main/packages/rpc-types#types).

**Comment by alexpyattaev** [comment]
If `Packet` lives in solo crate this solves nothing and creates more work for maintainers and the linker. What we need is to have ability to bump versions of all crates independently of agave versions. In this case, if we want to make changes to public API, we just roll the major version on e.g. solana-turbine and push on. Who cares if `agave-3.0.0` depends on `solana-turbine-45.17.2` ?

**Comment by t-nelson** [comment]
`Packet` should not exist

**Comment by steviez** [comment]
The latest leaning has been to put items behind a `agave-unstable-api` feature; the feature is non-default and the name implies that it is giving you types that we make no promises on.

Some stuff was put behind this feature for v3.0 which seems appropriate and some new stuff has come in behind that feature. However, I'm thinking we probably shouldn't move any existing stuff behind that feature until we do another major bump (ie v4.0) if we want to be proper.

We had been intending to move vast amount of the codebase before the v3.0 cut, but that didn't happen unfortunately and the v3.0.0 tag has already been cut

**Comment by t-nelson** [comment]
> However, I'm thinking we probably shouldn't move any existing stuff behind that feature until we do another major bump (ie v4.0) if we want to be proper.

i think we can move stuff now, but add the `agave-unstable-api` feature to default, then at next major all we have to do is drop default

**Comment by apfitzge** [comment]
Have we considered that our junk crates should just not be published at all? That way we don't need to worry about semver or not...it can exist as a concept within our repository, and we already have the name reserved on crates.io so no one can take it. But we don't don't need to continue to publish them.

Stuff like `solana-core` has very little use outside of our repository.

**Comment by t-nelson** [comment]
`solana-core` is a dependency of `agave-validator`, so we'd have to remove publish there too

**Comment by joncinque** [comment]
According to the docs at https://docs.anza.xyz/backwards-compatibility#rust-crates, we only care about:

* solana-program
* solana-sdk
* solana-client
* solana-cli-config
* agave-geyser-plugin-interface

Let's assume that all of those also include their internal dependencies, ie solana-address which is used by solana-program. If people complain about crates outside of that, then we can just point people to our documentation and say "that's our policy".

**Comment by t-nelson** [comment]
> If people complain about crates outside of that, then we can just point people to our documentation and say "that's our policy".

that is what we have been doing. it does not work. i do not want to hear from them at all

**Comment by joncinque** [comment]
> that is what we have been doing. it does not work. i do not want to hear from them at all

Gotcha, so the outcome for this issue is to stop complaints.

@sam0x17 had a really interesting idea, that involved using a macro to mark a function as deprecated if the feature isn't enabled. That'll cause a lot of churn in Agave, so maybe not the best option in the short term

Or another idea, how about marking the whole crate as deprecated if the feature is disabled?

**Comment by t-nelson** [comment]
if the feature is off, the code is removed, not deprecated. if we wanted to do that _right_ it would force us to do the reorganization around what if anything we _do_ want to commit to honoring semver for on each crates. which we will missing 4.0 with for sure

what problem are we trying to solve? we do not care what any consumers think about this action. our intentions do not change. the crate feature is purely to force them into admitting that they are willfully using unstable api. if we were smart, we never would have taken these internal crates out of v0

**Comment by t-nelson** [comment]
> Or another idea, how about marking the whole crate as deprecated if the feature is disabled?

i think i can get behind this. a little more noisy for the move to 4.0, the idea would be like this i assume? (atop #8424)
```patch
diff --git a/Cargo.toml b/Cargo.toml
index ae109fbbbc..96331cd215 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -414,7 +414,7 @@ solana-compute-budget-interface = "3.0.0"
 solana-compute-budget-program = { path = "programs/compute-budget", version = "=3.1.0" }
 solana-config-interface = "2.0.0"
 solana-connection-cache = { path = "connection-cache", version = "=3.1.0", default-features = false, features = ["agave-unstable-api"] }
-solana-core = { path = "core", version = "=3.1.0" }
+solana-core = { path = "core", version = "=3.1.0", features = ["agave-unstable-api"] }
 solana-cost-model = { path = "cost-model", version = "=3.1.0" }
 solana-cpi = "3.0.0"
 solana-curve25519 = { path = "curves/curve25519", version = "=3.1.0" }
diff --git a/core/Cargo.toml b/core/Cargo.toml
index 877113f78b..cb1ddf323d 100644
--- a/core/Cargo.toml
+++ b/core/Cargo.toml
@@ -13,7 +13,6 @@ edition = { workspace = true }
 targets = ["x86_64-unknown-linux-gnu"]

 [features]
-default = ["agave-unstable-api"]
 agave-unstable-api = []
 dev-context-only-utils = [
     "solana-perf/dev-context-only-utils",
diff --git a/core/src/lib.rs b/core/src/lib.rs
index d6d904c2cf..54f04ffece 100644
--- a/core/src/lib.rs
+++ b/core/src/lib.rs
@@ -1,4 +1,11 @@
-#![cfg(feature = "agave-unstable-api")]
+#![cfg_attr(
+    not(feature = "agave-unstable-api"),
+    deprecated(
+        since = "3.1.0",
+        note = "This crate has been marked for formal inclusion in the Agave Unstable \
+        API. From v4.0.0 onward, the `agave-unstable-api` crate feature must be specified \
+        to continue its use"
+))]
 #![cfg_attr(feature = "frozen-abi", feature(min_specialization))]
 #![allow(clippy::arithmetic_side_effects)]
 #![recursion_limit = "2048"]
```

**Comment by joncinque** [comment]
Yeah exactly. I think it gets the point across clearly to downstream devs, without being as destructive as removing all the code.

You can even go one step further with that message and say "By enabling the `agave-unstable-api` feature, you are willfully using an unstable API that may be broken at any time."

**Comment by t-nelson** [comment]
🫡 
```patch
-        to continue its use"
+        to acknowledge use of an interface that may break without warning."
```

**Comment by sam0x17** [comment]
Yeah my thought was just that deprecation warnings for 3-6 months should precede anything drastic. Warnings and compiler errors are ultimately the _only_ way to get most ecosystem users to notice anything

**Comment by steviez** [comment]
> According to the docs at ...

Slight aside but we should probably update those docs to say "the crates within sdk repo" now that SDK components have been broken out
